src/python/pants/bsp/rules.py

```diff
@@ -1,10 +1,34 @@
 # Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
-from pants.bsp.spec import BuildServerCapabilities, InitializeBuildParams, InitializeBuildResult
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+from pants.bsp.spec import (
+    BuildServerCapabilities,
+    BuildTarget,
+    InitializeBuildParams,
+    InitializeBuildResult,
+    WorkspaceBuildTargetsParams,
+    WorkspaceBuildTargetsResult,
+)
+from pants.engine.internals.selectors import Get, MultiGet
 from pants.engine.rules import QueryRule, collect_rules, rule
+from pants.engine.unions import UnionMembership, union
 from pants.version import VERSION
 
 
+@union
+class BSPBuildTargetsRequest:
+    """Request language backends to provide BSP `BuildTarget` instances for their managed target
+    types."""
+
+
+@dataclass(frozen=True)
+class BSPBuildTargets:
+    targets: tuple[BuildTarget, ...]
```

Comment:
Not a blocker, but: it sortof seems like 95% of the fields of `BuildTarget` could be computed generically, and only `BuildTarget.{data,data_kind}` are necessarily backend-specific? If that's the case, this `@union` should maybe be for a much smaller request to fill in those fields.

Reply:
I agree which is why I intend for the BSP core rules to expose a helper function that the individual language backends can use to fill in a BSP `BuildTarget`. Similar to how file-level target generation uses a helper method invoked by implementations of `GenerateTargetsRequest`.

Reply:
Yeeea, but I'm working on deleting/privatizing that helper function in #14377, and it's been a bear. The trouble with helper functions is that their usage can diverge, since both the inputs and outputs can be modified. Keeping an API bounded (in terms of which fields you can actually affect) from the start and expanding it only as more features are needed would be my preference here.

Reply:
The main issue is that I don't know yet what a "good" API will look like. All I'm trying for now is enough scaffolding so I can iterate on Scala support over BSP, which for now means doing that work in the Scala backend. I would not be surprised if I am able to refactor out some bits back into the core BSP code later, but that is not now.

Reply:
It is much easier to expand a declarative API than it is to contract an imperative API. It's similar to premature generalization to give "all the power" from the get go.